File: python/maze/Grid.py
import random
from maze.Cell import Cell


class Grid:
    """
    Grid is a wrapper object around a two-dimensional list of Cells.
    Grid's are passed to maze generation algorithms, which will then carve passages between cells, using the cell's
     link() method.
    """

    def __init__(self, rows: int, cols: int):
        self.rows = rows
        self.cols = cols
        # build 2D grid with default cell values
        self.grid = [[Cell(r, c) for c in range(cols)] for r in range(rows)]
        self._build_cell_neighbors()

    def size(self):
        """return the dimension of this grid, which is  grid.rows * grid.cols"""
        return self.rows * self.cols

    def get(self, row: int, col: int) -> Cell:
        return self.grid[row][col]

    def random_cell(self):
        """returns a random cell in this grid"""
        return random.choice(random.choice(self.grid))

    def dead_ends(self) -> list:
        """
        returns a list of cells in this grid that are dead end cells. Dead-end cells only have one link
        into/out of them
        """
        return [cell for cell in self.cell_iterator() if len(cell.linked_cells()) == 1]

    def braid(self, p: float):
        """
        Adds braids to this maze by removing dead-end cells and turning them into loops
        :param p: a float value where 0.0 <= p <= 1.0, that is the percentage amount of dead-ends to remove.
        1.0 = remove all dead-ends, while a value of 0.5 would remove 50 percent of dead-ends
        :return: None, the grid is mutated in-place
        """
        # get all dead-ends in the grid and shuffle them
        dead_ends = self.dead_ends()
        random.shuffle(dead_ends)

        for cell in dead_ends:
            if len(cell.linked_cells()) == 1 and random.random() <= p:
                # get neighbors that are not linked to cell
                neighbors = list(filter(lambda c: not cell.is_linked(c), cell.neighbors()))

                # try to find neighbors that are also dead-ends, if possible
                best = list(filter(lambda c: len(c.linked_cells()) == 1, neighbors))

                # if no best cells could be found, then just use neighbors
                if not best:
                    best = neighbors

                # choose a random neighbor and link to it
                cell.link(random.choice(best))

    def _build_cell_neighbors(self):
        """sets the neighbor cells for each cell in the grid"""
        for r in range(self.rows):
            for c in range(self.cols):
                # cell will have a north neighbor
                if r > 0:
                    self.grid[r][c].north = self.grid[r - 1][c]
                # cell will have a south neighbor
                if r < self.rows - 1:
                    self.grid[r][c].south = self.grid[r + 1][c]
                # cell will have east neighbor
                if c < self.cols - 1:
                    self.grid[r][c].east = self.grid[r][c + 1]
                # cell will have west neighbor
                if c > 0:
                    self.grid[r][c].west = self.grid[r][c - 1]

    def row_iterator(self):
        """iterates over the rows of this grid"""
        for r in range(self.rows):
            yield self.grid[r]

    def cell_iterator(self):
        """iterates over the cells of this grid in row major order"""
        for r in range(self.rows):
            for c in range(self.cols):
                yield self.grid[r][c]

    def print_grid(self, distances=None) -> str:
        """
        pretty prints the cells and walls of the grid into a String. if distances is passed,then the cells
        distance information will be printed inside the cell body as a hexadecimal number
        """

        def cell_body(cell, distances):
            if distances and cell in distances.cells:
                return "{:02x}".format(distances.cells[cell])
            else:
                return "  "

        # build the top wall
        out = "+" + ("---+" * self.cols) + "\n"

        for row in self.row_iterator():
            top = "|"
            bottom = "+"

            # for each cell only need to check if an east or a south wall should be drawn
            for cell in row:
                # body contains what should be printed within a cell of the grid
                body = cell_body(cell, distances)

                # determine if eastern wall should be drawn
                if cell.east and cell.is_linked(cell.east):
                    # don't draw a east wall
                    top += f"{body}  "
                else:
                    # draw the east wall
                    top += f"{body} |"

                # determine if south wall should be drawn
                if cell.south and cell.is_linked(cell.south):
                    # don't draw a south wall
                    bottom += "   +"
                else:
                    # draw the south wall
                    bottom += "---+"

            out += top + "\n"
            out += bottom + "\n"

        return out

    def __str__(self) -> str:
        return self.print_grid()

File: java/src/main/java/com/mazes/Grid.java
package com.mazes;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Grid is a wrapper object around a two-dimensional "grid" of Cells.
 * Grid's are used by maze generation algorithms to hold the "carved" passages between cells. Passages
 * are created by calling the cell's link() method.
 */
public class Grid implements Iterable<Cell> {

    // total number of rows in this grid
    public final int rows;

    // total number of columns in this grid
    public final int cols;

    // holds the cells of this grid in a two-dimensional array
    Cell [][] grid;

    // random number generator
    Random random;

    /**
     * builds a new Grid containing the specified rows and columns.
     * @param rows the number of rows the grid should contain (i.e. its height)
     * @param cols the number of columns the grid should contain (i.e. its width)
     */
    public Grid(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.grid = buildCells(rows, cols);
        this.random = new Random();
    }


    /**
     *
     * @return the dimension of this grid, i.e. rows * cols
     */
    public int size() {
        return this.rows * this.cols;
    }


    /**
     *
     * @param row row index into the grid
     * @param col column index into the grid
     * @return the Cell at the specified row and col.
     * @throws IndexOutOfBoundsException if the provided row or col is out of bounds of this Grid
     */
    public Cell get(int row, int col) {
        return this.grid[row][col];
    }

    /**
     *
     * @return a random cell in this Grid
     */
    public Cell randomCell() {
        int row = random.nextInt(this.rows);
        int col = random.nextInt(this.cols);
        return this.grid[row][col];
    }

    /**
     * get dead-end cells
     * @return a List containing all the cells of this grid that are dead-ends. Dead-ends are cells that
     * only have one link into/out of them.
     */
    public List<Cell> deadEnds() {
        return Arrays.stream(this.grid)
                .flatMap(Arrays::stream)
                .filter(cell -> cell.links().size() == 1)
                .collect(Collectors.toList());
    }

    /**
     * Adds braids to this maze by removing dead-end cells and turning them into loops
     *
     * @param p a float value where 0.0 <= p <= 1.0, that is the percentage amount of dead-ends to remove.
     *          1.0 = remove all dead-ends, while a value of 0.5 would remove 50 percent of dead-ends
     */
    public void braid(float p) {
        List<Cell> deadEnds = this.deadEnds();
        Collections.shuffle(deadEnds);

        for (Cell cell: deadEnds) {
            if (cell.links().size() == 1 && random.nextFloat() <= p) {
                // get neighbors that are NOT linked to the current pos
                List<Cell> neighbors = cell.neighbors()
                        .stream()
                        .filter(nbr -> !cell.is_linked(nbr))
                        .collect(Collectors.toList());

                // try to find neighbors that are also a dead end, if possible
                List<Cell> best = neighbors.stream()
                        .filter(nbr -> nbr.links().size() == 1)
                        .collect(Collectors.toList());

                // otherwise just use the original neigbors list
                if (best.isEmpty()) {
                    best = neighbors;
                }

                // choose a random neighbor and link to it
                Cell randNbr = best.get(random.nextInt(best.size()));
                cell.link(randNbr, true);
            }
        }
    }

    /**
     *
     * @return an iterator of Cell[], over the rows of this grid
     */
    public Iterator<Cell[]> row_iterator() {
        return Arrays.stream(this.grid).iterator();
    }



    /**
     * builds the cells of this grid and sets the neighbors for each cell
     * @param rows the number of rows in this grid
     * @param cols the number of columns in this grid
     * @return a two dimensional array of Cells, with neighbor cells created for each cell
     */
    private Cell[][] buildCells(int rows, int cols) {
        Cell [][] grid = new Cell[rows][cols];

        // build the Cells of this grid, without any links to neighbors
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                grid[r][c] = new Cell(r, c);
            }
        }
        // now set the links to neighboring Cells for each cell
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                setNorthNeighbor(grid, r, c);
                setSouthNeighbor(grid, r, c);
                setEastNeighbor(grid, r, c);
                setWestNeighbor(grid, r, c);
            }
        }

        return grid;
    }

    // set the north neighbor for the cell at index row,col in the grid
    private void setNorthNeighbor(Cell[][] grid, int row, int col) {
        if (row > 0) {
            grid[row][col].setNorth(Optional.of(grid[row - 1][col]));
        }
    }

    // set the south neighbor for the cell at index row,col in the grid
    private void setSouthNeighbor(Cell[][] grid, int row, int col) {
        if (row < this.rows - 1) {
            grid[row][col].setSouth(Optional.of(grid[row + 1][col]));
        }
    }

    // set the eastern neighbor for the cell at index row,col in the grid
    private void setEastNeighbor(Cell[][] grid, int row, int col) {
        if (col < this.cols - 1) {
            grid[row][col].setEast(Optional.of(grid[row][col + 1]));
        }
    }

    // set the western neighbor for the cell at index row,col in the grid
    private void setWestNeighbor(Cell[][] grid, int row, int col) {
        if (col > 0) {
            grid[row][col].setWest(Optional.of(grid[row][col - 1]));
        }
    }

    /**
     * pretty prints the Grid into a String
     * @return a pretty printed version of the Grid suitable for display to STDOUT
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        // build the top wall of the grid
        sb.append("+").append("---+".repeat(this.cols)).append("\n");

        // iterate row by row and only check if an east wall needs to be drawn or a south wall
        Iterator<Cell[]> rowIter = this.row_iterator();
        while (rowIter.hasNext()) {
            StringBuilder top = new StringBuilder("|");
            StringBuilder bottom = new StringBuilder("+");

            Cell[] row = rowIter.next();
            for (Cell cell: row) {
                // determine if an eastern wall should be drawn
                if (cell.east.isPresent() && cell.is_linked(cell.east.get())) {
                    top.append("    "); // no east wall drawn
                }
                else {
                    top.append("   |"); // east wall is drawn
                }

                // determine if south wall should be drawn
                if (cell.south.isPresent() && cell.is_linked(cell.south.get())) {
                    bottom.append("   +");  // no south wall
                } else {
                    bottom.append("---+");  // draw a south wall
                }
            }
            sb.append(top).append("\n");
            sb.append(bottom).append("\n");
        }

        return sb.toString();
    }

    /**
     * pretty prints this Grid to a String, with the provided distance information printed into
     * the Cells of the grid.
     * @param distances a Distances object to display on this Grid
     * @return a String containing pretty-printed Grid
     */
    public String printDistances(Distances distances) {
        StringBuilder sb = new StringBuilder();
        // build the top wall of the grid
        sb.append("+").append("---+".repeat(this.cols)).append("\n");

        Iterator<Cell[]> rowIter = this.row_iterator();
        while (rowIter.hasNext()) {
            StringBuilder top = new StringBuilder("|");
            StringBuilder bottom = new StringBuilder("+");

            Cell[] row = rowIter.next();
            for (Cell cell: row) {
                // if the current cell is contained in the Distance object, print its distance info in the cell
                // as a hexa-decimal number, else print 2 spaces
                String body = distances.contains(cell) ? String.format("%2x", distances.get(cell)) : "  ";

                // determine if an eastern wall should be drawn
                if (cell.east.isPresent() && cell.is_linked(cell.east.get())) {
                    top.append(body).append("  "); // no east wall drawn
                }
                else {
                    top.append(body).append(" |"); // east wall is drawn
                }

                // determine if south wall should be drawn
                if (cell.south.isPresent() && cell.is_linked(cell.south.get())) {
                    bottom.append("   +");
                } else {
                    bottom.append("---+");
                }
            }
            sb.append(top).append("\n");
            sb.append(bottom).append("\n");
        }

        return sb.toString();
    }

    @Override
    public Iterator<Cell> iterator() {
        return new GridIterator();
    }

    /**
     * An iterator implementation that iterates over the Cells of a Grid in row order
     */
    class GridIterator implements Iterator<Cell> {
        // one-dimensional index into Grid cells array
        int idx = 0;

        @Override
        public boolean hasNext() {
            return idx < Grid.this.size();
        }

        @Override
        public Cell next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            // convert one-dimensional index into row,col indices
            int cur_row = idx / Grid.this.cols;
            int cur_col = idx % Grid.this.cols;
            idx++;
            return grid[cur_row][cur_col];
        }
    }

}
